python/whylogs/api/logger/experimental/logger/actor/process_rolling_logger_messages.py

```diff
@@ -0,0 +1,282 @@
+        if log_request is None:
+            return None
+
+        data["log_request"] = log_request
+        return data
+
+
+class PubSubEmbeddingDict(TypedDict):
+    subscription: str
+    message: PubSubMessage
+    log_embedding_request: LogEmbeddingRequestDict
+
+
+def _decode_pubsub_data(data: Union[PubSubEmbeddingDict, PubSubDict]) -> Optional[bytes]:
+    if "message" not in data or data["message"] is None:
+        _logger.error(f"Request missing message field {data}")
+        return None
+
+    message = data["message"]
+    encoded_data = message["data"]
+    return base64.b64decode(encoded_data)
+
+
+@dataclass
+class RawPubSubEmbeddingMessage:
+    request: bytes
+    request_time: int
+
+    def to_pubsub_embedding_message(self) -> Optional[PubSubEmbeddingDict]:
+        data: PubSubEmbeddingDict = orjson.loads(self.request)
+        log_request_dict_bytes = _decode_pubsub_data(data)
+
+        if log_request_dict_bytes is None:
+            return None
+
+        log_message = RawLogEmbeddingsMessage(request=log_request_dict_bytes, request_time=self.request_time)
+        log_request = log_message.to_log_embeddings_request_dict()
+        if log_request is None:
+            return None
+        data["log_embedding_request"] = log_request
+        return data
+
+
+@dataclass
+class RawLogEmbeddingsMessage:
+    request: bytes
+    request_time: int
+
+    def to_log_embeddings_request_dict(self) -> Optional[LogEmbeddingRequestDict]:
+        d: LogEmbeddingRequestDict = orjson.loads(self.request)
+        if "timestamp" not in d or d["timestamp"] is None:
+            d["timestamp"] = self.request_time
+
+        if "datasetId" not in d or d["datasetId"] is None:
+            _logger.error(f"Request missing dataset id {d}")
+            return None
+
+        if "embeddings" not in d or d["embeddings"] is None:
+            _logger.error(f"Request has no embeddings field {d}")
+            return None
+
+        if not isinstance(d["embeddings"], dict):
+            # TODO test recovering from errors like this. It seems to brick the container
+            _logger.error(
+                f'Expected a dictionary format for embeddings of the form {{"column_name": "embedding_2d_list"}}. Got {self.request}'
+            )
+            return None
+
+        return d
+
+
+def log_dict_to_data_frame(request: LogRequestDict) -> Tuple[pd.DataFrame, int]:
+    df = pd.DataFrame(request["multiple"]["data"], columns=request["multiple"]["columns"])
+    return df, len(df)
+
+
+def log_dict_to_embedding_matrix(request: LogEmbeddingRequestDict) -> Tuple[Dict[str, np.ndarray], int]:
+    row: Dict[str, np.ndarray] = {}
+    row_count = 0
+    for col, embeddings in request["embeddings"].items():
+        row[col] = np.array(embeddings)
+        row_count += len(embeddings)
+
+    return row, row_count
+
+
+def reduce_log_requests(acc: LogRequestDict, cur: LogRequestDict) -> LogRequestDict:
```

Comment:
With respect to function name: this is more like a 'concatenation' (or collect, or extend) than a 'reduce'.

Where is this function used? I'm trying to reason about the disjoint columns comment but I'm not sure I understand the scenario when this would be called.

Reply:
Two slightly separate things going on here.

First thing. This function is actually called all the time when logging data inside of the ProcessRollingLogger. Check `process_log_dicts`. Since the logger handles batches of data, the batches are "reduced" into a single data frame with `reduce_log_requests` literally as the reducer in the built in `reduce()` fn. That's what's going on in `_process_dicts` in the ProcessRollingLogger:

```python
                    giga_message = reduce(reducer, sub_group)  # Called giga because its big
```

Second thing. The disjoint column thing is referring to the following scenario.

```python
logger.log({'a': 1, 'b': 2})
logger.log({'b': 4, 'c': 3})
```

This is something that can happen within a single dataset. Some messages might only contain some columns of the dataset. That's fine normally, and I could probably even pass it to whylogs for logging too, but I can't always assume that I can "reduce" those two log messages down into a single row/data frame. Because of that, when I'm batching these things up, I make sub batches based on the things that I know can be merged into larger data frames before feeding them into whylogs. Those dimensions include time, column names, and dataset id.

All of that stuff is happening in ProcessRollingLogger's `_process_dicts` fn, which is basically where all the batching/logging happens. So, the messages are consumed in bulk from the actor's queue, and then they're batched up into the message type (logging message, status message, etc), and handled in bulk. That means doing data merging in the case of the log messages so minimize the whylogs calls/maximize the data per whylogs call.